File: sample/java/10_plasma_effect.java
final class _10_plasma_effect {
    private _10_plasma_effect() {
    }


    // 10: Sample that outputs a plasma effect as a GIF.

    public static void run_10_plasma_effect() {
        long w = 320L;
        long h = 240L;
        long frames_n = 216L;
        String out_path = "sample/out/10_plasma_effect.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long t = 0L; t < frames_n; t += 1L) {
            java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(w * h);
            for (long y = 0L; y < h; y += 1L) {
                long row_base = y * w;
                for (long x = 0L; x < w; x += 1L) {
                    long dx = x - 160L;
                    long dy = y - 120L;
                    double v = math.sin((((double)(x)) + ((double)(t)) * 2.0) * 0.045) + math.sin((((double)(y)) - ((double)(t)) * 1.2) * 0.05) + math.sin((((double)(x + y)) + ((double)(t)) * 1.7) * 0.03) + math.sin(math.sqrt(dx * dx + dy * dy) * 0.07 - ((double)(t)) * 0.18);
                    long c = PyRuntime.__pytra_int((v + 4.0) * (255.0 / 8.0));
                    if (((c) < (0L))) {
                        c = 0L;
                    }
                    if (((c) > (255L))) {
                        c = 255L;
                    }
                    frame.set((int)((((row_base + x) < 0L) ? (((long)(frame.size())) + (row_base + x)) : (row_base + x))), c);
                }
            }
            frames.add(PyRuntime.__pytra_bytearray(frame));
        }
        gif.save_gif(out_path, w, h, frames, gif.grayscale_palette(), 3L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(frames_n));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/julia/10_plasma_effect.jl
include(joinpath(@__DIR__, "built_in", "py_runtime.jl"))

include(joinpath(@__DIR__, "std", "math.jl"))
math = (ceil=ceil, cos=cos, e=e, exp=exp, fabs=fabs, floor=floor, log=log, log10=log10, pi=pi, pow=pow, sin=sin, sqrt=sqrt, tan=tan)
include(joinpath(@__DIR__, "std", "time.jl"))
include(joinpath(@__DIR__, "utils", "gif.jl"))

# 10: Sample that outputs a plasma effect as a GIF.

function run_10_plasma_effect()
    w = 320
    h = 240
    frames_n = 216
    out_path = "sample/out/10_plasma_effect.gif"
    
    start = perf_counter()
    frames = Any[]
    
    for t in 0:frames_n - 1
        frame = __pytra_bytearray((w * h))
        for y in 0:h - 1
            row_base = (y * w)
            for x in 0:w - 1
                dx = (x - 160)
                dy = (y - 120)
                v = (((math.sin(((x + (t * 2.0)) * 0.045)) + math.sin(((y - (t * 1.2)) * 0.05))) + math.sin((((x + y) + (t * 1.7)) * 0.03))) + math.sin(((math.sqrt(((dx * dx) + (dy * dy))) * 0.07) - (t * 0.18))))
                c = __pytra_int(((v + 4.0) * (255.0 / 8.0)))
                if (c < 0)
                    c = 0
                end
                if (c > 255)
                    c = 255
                end
                frame[__pytra_idx((row_base + x), length(frame))] = c
            end
        end
        push!(frames, __pytra_bytes(frame))
    end
    save_gif(out_path, w, h, frames, grayscale_palette(), 3, 0)
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
end


run_10_plasma_effect();
